modules/core/src/matmul.dispatch.cpp

```diff
@@ -158,14 +158,20 @@ static bool ocl_gemm( InputArray matA, InputArray matB, double alpha,
     int depth = matA.depth(), cn = matA.channels();
     int type = CV_MAKETYPE(depth, cn);
 
-    CV_Assert_N( type == matB.type(), (type == CV_32FC1 || type == CV_64FC1 || type == CV_32FC2 || type == CV_64FC2) );
+    CV_Assert_N( type == matB.type(),
+        (type == CV_32FC1 || type == CV_64FC1 || type == CV_16FC1 ||
+         type == CV_32FC2 || type == CV_64FC2 || type == CV_16FC2) );
```

Comment:
Does it really work?

- matA of 32F
- matB of 16F

Reply:
The assertion's first condition, `type == matB.type()`, ensures that the two matrices have the same type.